lib/limiter/reservation.go

```diff
@@ -0,0 +1,133 @@
+package limiter
+
+import (
+	"context"
+	"sync"
+	"tcplb/lib/core"
+)
+
+type ClientReservation struct {
+	c core.ClientID
+}
+
+// MaxReservationsExceeded is the error returned by ClientReserver
+// when an attempted reservation fails because the client has too
+// many reservations.
+var MaxReservationsExceeded error = maxReservationsExceededError{}
+
+type maxReservationsExceededError struct{}
+
+func (maxReservationsExceededError) Error() string { return "maximum client reservations exceeded" }
+
+// ClientReserver represents an object that can acquire and release
+// reservations for clients.
+//
+// Multiple goroutines may invoke methods on a ClientReserver simultaneously.
+type ClientReserver interface {
+
+	// TryReserve attempts to acquire a reservation for the given client.
+	// If the attempt succeeds, the reservation is returned with nil error.
+	// If the attempt fails because the client has exceeded the maximum number
+	// of reservations, the zero reservation and MaxReservationsExceeded error
+	// will be returned.
+	//
+	// Implementations of ClientReserver are discouraged from blocking until
+	// a reservation becomes available.
+	TryReserve(ctx context.Context, c core.ClientID) (ClientReservation, error)
+
+	// ReleaseReservation releases a reservation that was previously acquired
+	// by TryReserve.
+	ReleaseReservation(ctx context.Context, r ClientReservation) error
```

Comment:
```suggestion
	ReleaseReservation(ctx context.Context, r *ClientReservation) error
```

so you can change the reservation if you need to? Although, admittedly, there isn't much to change right now.